java/src/main/native/src/CudaJni.cpp

```diff
@@ -41,19 +41,23 @@ void set_cudf_device(int device) {
  * is using the same device.
  */
 void auto_set_device(JNIEnv *env) {
-  if (Cudf_device != cudaInvalidDeviceId) {
-    if (Thread_device != Cudf_device) {
-      cudaError_t cuda_status = cudaSetDevice(Cudf_device);
-      jni_cuda_check(env, cuda_status);
-      Thread_device = Cudf_device;
+  try {
+    if (Cudf_device != cudaInvalidDeviceId) {
+      if (Thread_device != Cudf_device) {
+        JNI_CUDA_CHECK(env, cudaSetDevice(Cudf_device));
+        Thread_device = Cudf_device;
+      }
     }
   }
+  CATCH_STD(env, );
```

Comment:
This is incorrect, as it will catch the thrown exception, post a JVM exception, and then _allow the calling code to continue as if no error occurred_.  The original code posted a JVM exception and then threw a native exception to prevent this, and that behavior should be preserved.

Reply:
Created a new catch marco `CATCH_CUDA_ERROR_AND_THROW` which operates in catch-and-halt manner.